### DISCUSSION TITLE
Which problem option to select?

### DISCUSSION THREAD

**Discussion by bmorlo** [PyVRP/PyVRP#593] [Q&A] [answered]
Hi,

my plan is to edit your pyvrp package (on Linux), build it, and then publish it on PyPI to use it also on e.g. a windows PC.
I just came across the option `--problem {cvrp, vrptw}` - see below. As I am not working with time windows, I was curious, how the cvrp solver differs from the vrptw in terms of performance. Would you recommend simply sticking to the vrptw solver in any case?

Thanks & cheers

```
morlocb@abcde:~/coding_dir/PyVRP_BM_load_balancing$ poetry run python build_extensions.py --help
usage: build_extensions [-h] [--build_dir BUILD_DIR] [--build_type {debug,debugoptimized,release}] [--problem {cvrp,vrptw}] [--clean] [--verbose] [--use_pgo] [--additional ...]

options:
  -h, --help            show this help message and exit
  --build_dir BUILD_DIR
                        Directory for Meson to use while building extensions.
  --build_type {debug,debugoptimized,release}
                        The type of build to provide. Defaults to release mode.
  --problem {cvrp,vrptw}
                        Which type of solver to compile. Defaults to 'vrptw'.
  --clean               Clean build and installation directories before building.
  --verbose             Whether to print more verbose compilation output.
  --use_pgo             Whether to enable profile-guided optimisation.
  --additional ...      Extra Meson configuration options (passed verbatim to Meson).
```

**Comment by N-Wouda** [comment, marked as answer]
> my plan is to edit your pyvrp package (on Linux), build it, and then publish it on PyPI to use it also on e.g. a windows PC.

You don't need to push to PyPI to get the required wheels out - you can download them once the GitHub Action has completed. For us, that's every time the CD runs, for example [here](https://github.com/PyVRP/PyVRP/actions/runs/9234905102).

> As I am not working with time windows, I was curious, how the cvrp solver differs from the vrptw in terms of performance. Would you recommend simply sticking to the vrptw solver in any case?

It should be a bit (up to 20-25%?) faster because the CVRP-specific build removes all time window computations from the algorithm. This build option will probably be removed as part of https://github.com/PyVRP/PyVRP/issues/502, but we haven't gotten around to that yet. If you don't need anything duration related, this could be worthwhile. However, the actual performance differences in terms of solution quality will likely be fairly small.

**Reply by bmorlo** [reply]
Hi @N-Wouda,

allow me a few questions on your comment, as I'm quite new to working with github and PyPI.
> You don't need to push to PyPI to get the required wheels out - you can download them once the GitHub Action has completed. For us, that's every time the CD runs, for example here.

- What exactly is 'the GitHub Action' you are referring to?

- What is the CD / CD run?

- I've successfully pushed to PyPi from Linux using poetry. Unfortunately, it does not appear to be public yet. How did you guys manage to make it visible on PyPI? Anyhow, I can still call `pip install --upgrade pyvrp-load_balancing`. However, this fails (on Windows) when trying to build `wheel`. Do you have encountered something similar while working on your project?

Thanks again.

**Reply by N-Wouda** [reply]
Hi, sorry, this took a while.

>What exactly is 'the GitHub Action' you are referring to? What is the CD / CD run?

The thing we use to run our continuous integration (CI) and deployment (CD) pipelines. These pipelines are defined in workflows [here](https://github.com/PyVRP/PyVRP/tree/main/.github/workflows). They run every time we commit new code to the `main` branch, or open a PR with code to be merged into main. Basically these are checks that the new code is OK, and that the resulting PyVRP package still does everything we expect it to be able to do.

Part of that is to build a full package for each platform we support, and to check that these packages work. Since we have native extensions compiled from our C++ codebase, we produce binary packages. These are called 'wheels' in the Python world. Those wheels may be downloaded once the CD completes, and can be installed just like any other package can be (e.g., `pip install <path to wheel>` works just fine).

Your fork of PyVRP has the CD workflow, which you can run in your forked repository. You can thus produce new Windows versions of your fork of PyVRP really easily using our CD pipeline.

>I've successfully pushed to PyPi from Linux using poetry. Unfortunately, it does not appear to be public yet. How did you guys manage to make it visible on PyPI? Anyhow, I can still call pip install --upgrade pyvrp-load_balancing. However, this fails (on Windows) when trying to build wheel. Do you have encountered something similar while working on your project?

![image](https://github.com/PyVRP/PyVRP/assets/16272507/07519f86-bffc-4b56-b7bc-16bff974b646)

See above, which I got from [here](https://pypi.org/project/pyvrp-load_balancing/#files). You have uploaded a source distribution and a wheel for Linux (with Python 3.10). Thus, when you run `pip install pyvrp-load_balancing` on Windows, it will try to compile the sources since no wheel could be found that works on Windows. That compilation then fails. The solution is to also upload wheels for Windows - or bypass PyPI altogether and just use the wheels built by the CD directly.

**Reply by bmorlo** [reply]
Okay, thank you very much for the detailed answers. While I do not get everything just yet, it definitely helps with my understanding of your workflow and package management. Since I do not have too much time left, I will continue to run things solely on my Linux PC.
Cheers.

**Comment by leonlan** [comment]
Hi @bmorlo,

Let me try to steer you a bit in a different direction. You are asking a lot of questions - which is not a problem in itself - but it's about a lot of things that I don't think will help you with your current work. First some clarifying assumptions:
- I assume you are working on your thesis based on your first message to me on LinkedIn.
- I assume you are trying to build load-balancing into PyVRP for your thesis use case.

You need at least 2+ months (full-time) before you can modify the code of PyVRP to your liking, assuming you are already proficient in Python and C++. You can read the paper for details on the algorithmic and mathematical background. PyVRP implements a complex local search solver, and if you cannot modify that correctly, you are likely not able to achieve your goal of building a load-balancing feature.

For a master's thesis, adopting a state-of-the-art solver is often too much because of time constraints. Most master's theses implement much simpler (meta-)heuristics, which is perfectly fine. For example, I have seen many master students implement an adaptive large neighborhood search metaheuristic for VRPs, see, e.g., https://alns.readthedocs.io/en/latest/examples/capacitated_vehicle_routing_problem.html. ALNS is easy to understand and implement, and importantly, it works well if implemented correctly. Of course, I'm not your advisor, so you have to check with your own advisor to see if that fits the scope of your thesis.

If you find yourself new to GitHub, consider starting with smaller projects first. For instance, the ALNS package I mentioned is considerably simpler than PyVRP; it was actually the first project I contributed to. Read the documentation, use StackExchange, ask GPT questions, etc., to get yourself familiar with all the Git(hub) bells and whistles. Additionally, if you're interested in learning about Python packaging, I recommend checking out this [guide on creating a Python package](https://mathspp.com/blog/how-to-create-a-python-package-in-2022). Although it's from 2022, most of the information remains relevant and current.

I hope this advice helps you further in your thesis. Wishing you the best of luck, and feel free to reach out if you need more tips.

**Reply by bmorlo** [reply]
Hi @leonlan,

I really appreciate your remarks! Thank you so much for that! And yes, your assumptions are correct regarding my master's thesis and my current implementation goal.

Let me explain a few more things about me and my work (I have another two weeks to get the load balancing feature to work with pyvrp and I'm afraid I will have to try at least this week day and night):

- I studied Vidals' HGA quite a lot in the last weeks (reading the papers) and my thesis is set to evolve around that
- I've been working with the `pyvrp` package for the last two months
- I am fairly proficient in python and C++
- I haven't really worked with GitHub before (I'm already getting better, I have a good friend that offered to help me from now on)
- I haven't modified or created a python package before (I believe I can figure this out)

Allow me further to propose my approach regarding the implementation of a load balancing feature. It would be awesome to get you guys' feedback on that.

- I plan to stick very closely to your price collecting implementation.
- I was thinking to only modify the objective function and add new variables to the route class and the solution class. Using these variable (and some `find_maximum_capacity_gap` function) I believe I could already make sure I get a (much) worse objective value whenever there is a highly underutilized vehicle in the solution (a very low number of visited clients, relative to the other routes aka a large capacity gap)
- You mentioned I would have to modify the local search method. Would this still be the case here? I'm probably mistaken, but shouldn't the code still work just as good if I do not touch your granular neighborhood search?

**Reply by leonlan** [reply]
I respect your drive to work on this, and it's good to know that you're familiar with the literature & code, so let's make the best out of this then the next two weeks. :-)

> I was thinking to only modify the objective function and add new variables to the route class and the solution class. Using these variable (and some find_maximum_capacity_gap function) I believe I could already make sure I get a (much) worse objective value whenever there is a highly underutilized vehicle in the solution (a very low number of visited clients, relative to the other routes aka a large capacity gap)

That could work. I still don't know what "load-balancing" formally means in your context but if it's simply penalizing underutilized vehicle then you can try to mimic the principles of hwo ``loadPenalty`` is calculated. Instead of taking "capacity" you can introduce some "target capacity" which updates based on the average (?) route capacity utilization of your solution and add objective penalty based on that.

> You mentioned I would have to modify the local search method. Would this still be the case here? I'm probably mistaken, but shouldn't the code still work just as good if I do not touch your granular neighborhood search?

That's correct, I was wrong here. Only changing CostEvaluator could be enough.

> I haven't modified or created a python package before (I believe I can figure this out)

Maybe I understand it wrong but why do you need to make a Python package? If you can run it in your Linux environment, why not continue there? If you want to make it work for Windows you need to build it specifically for Windows using a compatible toolchain. You can check out our [CD](https://github.com/PyVRP/PyVRP/blob/main/.github/workflows/CD.yml) to see how we do that with Github. (This is also the answer to your earlier questions on the Github Action & CD run.)

**Reply by bmorlo** [reply]
- Load balancing (as far as I know) simply is the additional objective to have each vehicle cover a similar amount of demand. In my case, as I'm working with `delivery=1` for all clients, this gets me to the "makespan minimization" that I was thinking about two weeks ago. I would like to actively optimize for tours of similar lengths in the solutions.

- You're right. I can definitely deprioritize the package for now. I will try to get things implemented by Wednesday. I'll keep you updated. Thanks again for the support 😎

**Reply by bmorlo** [reply]
Hi @leonlan,
I am happy to report that I was able to make the necessary modifications in `CostEvaluator.h` , `Solution.h`, `Solution.cpp` to run my forked repository as a pyvrp_load_balancing package, see [here](https://github.com/bmorlo/PyVRP_load_balancing). Everything is working as expected, as far as I can tell.

- A short summary of what I did: I stuck very close to your code lines for price collecting. Now, whenever the CostEvaluator evaluates a solution, a very high penalty term gets added to the objective function. This penalty term is: $penalty = a * \max_k (vehicleCapacity_k - numberVisits_k)$, where `k` is the vehicle index and `a` is a scaling parameter. I currently simply set `a` to e.g. 10000 for my current test instance).

I now have a few questions:

- Whenever I push a modification to my forked repository, I get the message `CD: Build wheels on ubuntu-22.04 failed`. Is this simply because your checks detect something that should not be there or because there is a fundamental problem with what I added?

Some time ago @N-Wouda explained to me regarding your price collecting penalties: _To make the distance worth travelling the prizes should be high enough. It really depends on your application what "high enough" entails exactly. In a problem I'm currently working on, my prizes relate to a probability estimate of some event happening at each client location. The estimate is in [0, 1]. **I scale that estimate by a parameter `b` to balance it with the distance estimates. Finding the right value of `b` requires some tuning on the topology of the problem.**_
- Now my question on this would be: How do I actually go about determining the right value of my scaling parameter `a` in my load balancing approach? Right now I am simply using a very high value. However, I would like to leave the solver with some degrees of freedom to be able to find a best solution even with slightly unequal routes. I was thinking about something like $a = \max_{i,j} (edgeDistance_{i,j})$. As I am currently working through the literature anyway, I will try to find something on that. But I thought I might as well ask you guys how you are actually getting to those fine-tuned values in your problems.

**Comment by leonlan** [comment]
> A short summary of what I did: I stuck very close to your code lines for price collecting. Now, whenever the CostEvaluator evaluates a solution, a very high penalty term gets added to the objective function. This penalty term is: , where k is the vehicle index and a is a scaling parameter. I currently simply set a to e.g. 10000 for my current test instance).

This kind of works, but the algorithm is not complete. You still need to change a few things in `CostEvaluator` to have the correct behavior HGS. 

> Whenever I push a modification to my forked repository, I get the message CD: Build wheels on ubuntu-22.04 failed. Is this simply because your checks detect something that should not be there or because there is a fundamental problem with what I added?

CI/CDs fail simply because the checks in place do not pass. See https://github.com/bmorlo/PyVRP_load_balancing/actions for the details.

> Now my question on this would be: How do I actually go about determining the right value of my scaling parameter a in my load balancing approach? Right now I am simply using a very high value. However, I would like to leave the solver with some degrees of freedom to be able to find a best solution even with slightly unequal routes. I was thinking about something like . As I am currently working through the literature anyway, I will try to find something on that. But I thought I might as well ask you guys how you are actually getting to those fine-tuned values in your problems.

That's a matter of parameter tuning, I cannot say what works and what does not work for your problem.

**Reply by bmorlo** [reply]
>This kind of works, but the algorithm is not complete. You still need to change a few things in CostEvaluator to have the correct behavior HGS.

What exactly would you say is it that I need to change? I did make some adjustments to CostEvaluator. Do you mean I have to make some more adjustments?

[My CostEvaluator](https://github.com/bmorlo/PyVRP_load_balancing/blob/main/pyvrp%2Fcpp%2FCostEvaluator.h)

**Reply by N-Wouda** [reply]
Almost certainly in `CostEvaluator::deltaCost`, which is used in the local search. `CostEvaluator::penalisedCost` appears to be modified, but that's not used by the local search at all. That's only used to evaluate complete `Solution` and `Solution::Route` objects - we use something slightly different inside the local search.

**Reply by bmorlo** [reply]
Hmm. Thank you for the answer. I understand. To an extent. Allow me a few more questions:

I only modified the part of `CostEvaluator` which evaluates a `Solution`, (and not a single `Route`) via the `penalisedCost` function. The reason for that is that I believed I could simply penalise the solutions that are "bad" in my context and that the local search would not have to change at all.

- Do I now understand it correctly, that I additionally need to modify `deltaCost` such that the local search can actively improve a solution which is "bad" in my context to a, in my context, "better" solution? Stated differently, I need to tell the local search what kind of solutions I want. This, I'm currently not doing. But this is critical for the performance.

- You did not incorporate any (uncollected) prizes in the `deltaCost` function. Doesn't this mean that you are **not** actively searching for "better-prized" solutions during the local search?

**Reply by N-Wouda** [reply]
In order: 
- Yes, probably; 
- That's handled elsewhere in the local search. I suggest to read the code to get the full details.

**Reply by bmorlo** [reply]
I believe I implemented things properly this time. The best found solutions now look much better in terms of distance. At least compared to when I last reported "success".

I completely changed my approach towards the one that @leonlan suggested in the beginning of the week: Mimicking the `loadPenalty()` function in `CostEvaluator.h` by adding an `underUtilizationPenalty()` function which penalises "underutilized" routes and solutions. I also additionally modified `CostEvaluator::deltaCost` in the way that @N-Wouda recommended. I am currently using a very high value (10000) to penalize.

My modified `CostEvaluator.h` file can be found [here](https://github.com/bmorlo/PyVRP_load_balancing/blob/Version_2_loadPenalty/pyvrp/cpp/CostEvaluator.h) - I marked my changes with `// @bmorlo`.

Again, allow me a few questions to make sure I'm not producing complete garbage:
-  In `CostEvaluator::deltaCost` I did not quite understand what `skipLoad` does and why it is important. Should I put my code lines inside the `if` statements, too (see below)?
```
if constexpr (!skipLoad)
    {
        out -= loadPenalty(route->load(), route->capacity());
    }
// @bmorlo
out -= underUtilizationPenalty(route->load(), route->capacity());
```

- I noticed that `loadPenalty()` was called also in `SwapStar.cpp`. So far it's not quite clear to me what is being done with it there (see code below). Do I have to use my `underUtilizationPenalty()` function here too? I am only working with deliveries in my work.
```
Cost SwapStar::evaluate(Route *routeU, Route *routeV, CostEvaluator const &costEvaluator)
{
    best = {};

    if (updated[routeU->idx()])
        updateRemovalCosts(routeU, costEvaluator);

    if (updated[routeV->idx()])
        updateRemovalCosts(routeV, costEvaluator);

    for (auto *U : *routeU)
        for (auto *V : *routeV)
        {
            // The following lines compute a delta cost of removing U and V from
            // their own routes and inserting them into the other's route in the
            // best place. This is approximate since removal and insertion are
            // evaluated separately, not taking into account that while U leaves
            // its route, V will be inserted (and vice versa).
            Cost deltaCost = 0;

            // Separating removal and insertion means that the effects on load
            // are not counted correctly: during insert, U is still in the
            // route, and now V is added as well. The following addresses this
            // issue with an approximation, which is inexact when there are both
            // pickups and deliveries in the data. We do not evaluate load when
            // calculating remove and insert costs - that is all handled here.
            // So it's pretty rough but fast and seems to work well enough for
            // most instances.
            ProblemData::Client const &uClient = data.location(U->client());
            ProblemDataSwapStar::Client const &vClient = data.location(V->client());
            auto const uLoad = std::max(uClient.delivery, uClient.pickup);
            auto const vLoad = std::max(vClient.delivery, vClient.pickup);
            auto const loadDiff = uLoad - vLoad;

            deltaCost += costEvaluator.loadPenalty(routeU->load() - loadDiff,
                                                   routeU->capacity());
            deltaCost -= costEvaluator.loadPenalty(routeU->load(),
                                                   routeU->capacity());

            deltaCost += costEvaluator.loadPenalty(routeV->load() + loadDiff,
                                                   routeV->capacity());
            deltaCost -= costEvaluator.loadPenalty(routeV->load(),
                                                   routeV->capacity());

            deltaCost += removalCosts(routeU->idx(), U->client());
            deltaCost += removalCosts(routeV->idx(), V->client());

            auto [extraV, UAfter] = getBestInsertPoint(U, V, costEvaluator);
            deltaCost += extraV;

            if (deltaCost >= 0)  // continuing here avoids evaluating another
                continue;        // costly insertion point below

            auto [extraU, VAfter] = getBestInsertPoint(V, U, costEvaluator);
            deltaCost += extraU;

            if (deltaCost < best.cost)
            {
                best.cost = deltaCost;

                best.U = U;
                best.UAfter = UAfter;

                best.V = V;
                best.VAfter = VAfter;
            }
        }

    // It is possible for positive delta costs to turn negative when we do an
    // exact evaluation. But in practice that almost never happens, and is not
    // worth spending time on.
    if (best.cost >= 0)
        return best.cost;

    return evaluateMove(best.V, best.VAfter, best.U, costEvaluator)
           + evaluateMove(best.U, best.UAfter, best.V, costEvaluator);
}
```
Thanks & cheers as always.

**Reply by N-Wouda** [reply]
I don't know. Maybe try both approaches and see what works best?

**Reply by leonlan** [reply]
> In CostEvaluator::deltaCost I did not quite understand what skipLoad does and why it is important. Should I put my code lines inside the if statements, too (see below)?

`skipLoad` skips the load calculation for the delta cost evaluation. Whether that's useful or not for your case depends on whether you have cases where you should skip it or not.

>   I noticed that loadPenalty() was called also in SwapStar.cpp. So far it's not quite clear to me what is being done with it there (see code below). Do I have to use my underUtilizationPenalty() function here too? I am only working with deliveries in my work.

I suggest reading Vidal's 2022 paper on HGS-CVRP to better understand SwapStar. I don't know if you should use your function here, but probably yes.

**Reply by bmorlo** [reply]
> skipLoad skips the load calculation for the delta cost evaluation. Whether that's useful or not for your case depends on whether you have cases where you should skip it or not.

I currently cannot think of any cases where you would skip the load calculation in the delta cost. This would essentially mean one would not penalizing excess load, right? Maybe you have an example in mind that makes this clearer for me @leonlan

**Reply by leonlan** [reply]
SwapStar has many places where load calculations are skipped. The code snippet that you referenced includes an explanation.

**Reply by bmorlo** [reply]
Right, thanks for the hint. Now things make sense. I will adjust my fork. Will report back tomorrow. I believe then this discussion can be closed. Cheers

**Reply by bmorlo** [reply]
Things still work well after I have modified `SwapStar`. However: I noticed that I'm getting only feasible solutions (this was also the case before I modified `SwapStar`) when using very high penalties for "underutilized" solutions and routes. From my understanding this is fine and to be expected. When you penalize "underutilized" routes as heavily as I do, the chances of getting a (capacity violating) infeasible solution are very low.

My question now is: Does only getting feasible solutions impact the performance of the solver from your point of view? I'm unsure here. 
- On the one hand I understand that it is important to also explore infeasible (in terms of `excessLoad`) solutions to find the best solutions since these are most likely close to infeasibility.
- However, in my case, I am interested in solutions that exhibit a minimum load in any of their routes which is as close to the capacity limit as possible. This balances the load between routes since underutilized routes get a penalty. I believe it is only natural that no infeasible solutions appear when I'm penalizing quite strictly in my scenario. Indeed if I decrease the penalty, I get also infeasible solutions. With this lower penalty, the best solution exhibits a more uneven vehicle utilization but has a better objective value.

**Reply by leonlan** [reply]
> Does only getting feasible solutions impact the performance of the solver from your point of view?

Yes, which is precisely the reason why we have infeasible solutions. I cannot comment on your case whether it matters or not. You have to run your own benchmarks to figure it out.

**Comment by bmorlo** [comment]
Alright. I benchmarked the latest version of my code on one of your test sets (which I modified to set a fixed number of vehicles and also have unit-demand per client) against your original PyVRP version. The performance with very even tours is only about 1%  worse. See the two screenshots below (unfortunately I cut off the tours... so you gotta believe me her xD). The printed progress and the time stamps look good too. (If you see something that is off please let me know!) This should more than work for my thesis.

I'll leave the link to my final repository [here](https://github.com/bmorlo/PyVRP_LB_new)
Cheers, Benedikt.

1) PyVPR-standard:
![Standard](https://github.com/PyVRP/PyVRP/assets/164156189/d5e4a5de-7d8a-4f8d-9f7b-fe4197ec228a)

2) Load balancing:
![LoadB](https://github.com/PyVRP/PyVRP/assets/164156189/322affc0-19b9-4eb5-875f-d5f469f6b381)

**Reply by N-Wouda** [reply]
Great!
